File: Problem_Set_2/Vigenere/vigenere.c
// Implement the vigenere cipher
#include <cs50.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>

//Important functions prototypes
bool if_all_alpha(string text);
void shift(string keyword, int keywordInts[]);
string encipher(string plaintext, int count, int keywordInts[]);

int main(int argc, string argv[])
{
    if (argc != 2)
    {
        printf("Usage: ./vigenere keyword\n");
        return 1;
    }
    else
    {
        if (!if_all_alpha(argv[1]))
        {
            printf("Usage: ./vigenere keyword\n");
            return 1;
        }
        else
        {
            int length = strlen(argv[1]), keywordInts[length];
            shift(argv[1], keywordInts);
            string plaintext = get_string("plaintext: ");
            string ciphertext = encipher(plaintext, length, keywordInts);
            printf("ciphertext: %s\n", ciphertext);
        }
    }
    return 0;
}

//check if all the characters in the keyword are alphabets
bool if_all_alpha(string text)
{
    int alphabet = 0, n = strlen(text);
    for (int i = 0; i < n; i++)
    {
        if (isalpha(text[i]))
        {
            alphabet++;
        }
    }
    return alphabet == n;
}

//convert each character of the keywork to its appropriate integer value
void shift(string keyword, int keywordInts[])
{
    for (int i = 0, n = strlen(keyword); i < n; i++)
    {
        char c = keyword[i];
        if (islower(c))
        {
            keywordInts[i] = c - 'a';
        }
        else if (isupper(c))
        {
            keywordInts[i] = c - 'A';
        }
    }
}

//
string encipher(string plaintext, int count, int keywordInts[])
{
    string ciphertext = plaintext;
    int j = 0;
    for (int i = 0, n = strlen(plaintext); i < n; i++)
    {
        if (isalpha(plaintext[i]) != 0)
        {
            char ch = plaintext[i];
            // Drag down the character ascii value in the range of 0-26
            // then apply the cipher and get the character back in the right range
            // depending of wether it's an uppercase or lowercase character
            if (islower(ch))
            {
                ciphertext[i] = (((ch - 97) + keywordInts[(j % count)]) % 26) + 97;
            }
            else if (isupper(ch))
            {
                ciphertext[i] = (((ch - 65) + keywordInts[(j % count)]) % 26) + 65;
            }
            j++;
        }
        else
        {
            ciphertext[i] = plaintext[i];
        }
    }
    return ciphertext;
}

File: Problem_Set_6/vigenere/vigenere.py
# Implement the vigenere cipher


from cs50 import *


def main():
    if (len(sys.argv) != 2):
        print("Usage: python vigenere.py k")
        sys.exit(1)
    else:
        if (not sys.argv[1].isalpha()):
            print("Usage: python vigenere.py k")
            sys.exit(1)
        else:
            plaintext = get_string("plaintext: ")
            keyC = correspond(sys.argv[1])
            ciphertext = encipher(plaintext, keyC)
            print(f"ciphertext: {ciphertext}")


def correspond(keyword):
    keyC = []
    for i in keyword:
        if i.isupper():
            keyC.append(ord(i) - 65)
        elif i.islower():
            keyC.append(ord(i) - 97)
    return keyC


# Enciphering the given string


def encipher(plaintext, keyC):
    ciphertext = ""

    j = 0
    for i in range(len(plaintext)):
        if (plaintext[i].isalpha()):
            if (plaintext[i].islower()):
                ciphertext += chr((((ord(plaintext[i]) - 97) + keyC[(j % len(keyC))]) % 26) + 97)
            elif (plaintext[i].isupper()):
                ciphertext += chr((((ord(plaintext[i]) - 65) + keyC[(j % len(keyC))]) % 26) + 65)
            j += 1
        else:
            ciphertext += plaintext[i]

    return ciphertext


if __name__ == "__main__":
    main()
